dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,387 @@
+{
+    CoerseValue();
+}
+
+void NumberBox::OnMaximumPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    CoerseValue();
+}
+
+void NumberBox::OnNumberFormatterPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    // Update text with new formatting
+    UpdateTextToValue();
+}
+
+void NumberBox::ValidateNumberFormatter(winrt::INumberFormatter2 value)
+{
+    // NumberFormatter also needs to be an INumberParser
+    if (!value.try_as<winrt::INumberParser>())
+    {
+        throw winrt::hresult_error(E_INVALIDARG);
+    }
+}
+
+void NumberBox::OnSpinButtonPlacementModePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    SetSpinButtonVisualState();
+}
+
+void NumberBox::OnTextPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    if (auto textBox = m_textBox.get())
+    {
+        textBox.Text(Text());
+    }
+}
+
+void NumberBox::OnBasicValidationModePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    ValidateInput();
+}
+
+void NumberBox::OnTextBoxLostFocus(winrt::IInspectable const& sender, winrt::RoutedEventArgs const& args)
+{
+    ValidateInput();
+}
+
+void NumberBox::CoerseValue()
+{
+    // Validate that the value is in bounds
+    auto value = Value();
+    if (!IsInBounds(value) && BasicValidationMode() == winrt::NumberBoxBasicValidationMode::InvalidInputOverwritten)
+    {
+        // Coerse value to be within range
+        if (value > Maximum())
+        {
+            Value(Maximum());
+        }
+        else
+        {
+            Value(Minimum());
+        }
+    }
+}
+
+void NumberBox::ValidateInput()
+{
+    // Validate the content of the inner textbox
+    if (auto textBox = m_textBox.get())
+    {
+        auto text = textBox.Text();
+        
+        // Handles empty TextBox case, set text to current value
+        if (text.empty())
+        {
+            UpdateTextToValue();
+        }
+        else
+        {
+            if (AcceptsCalculation() && IsFormulaic(text))
+            {
+                NormalizeShorthandOperations();
+                EvaluateInputCalculation();
+            }
+
+            // Setting NumberFormatter to something that isn't an INumberParser will throw an exception, so this should be safe
+            auto numberParser = NumberFormatter().as<winrt::INumberParser>();
+            auto parsedNum = numberParser.ParseDouble(text);
+
+            if (!parsedNum)
+            {
+                if (BasicValidationMode() == winrt::NumberBoxBasicValidationMode::InvalidInputOverwritten)
+                {
+                    // Override text to current value
+                    UpdateTextToValue();
+                }
+            }
+            else
+            {
+                Value(parsedNum.Value());
+            }
+        }
+    }
+}
+
+void NumberBox::OnSpinDownClick(winrt::IInspectable const&  sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(false);
```

Comment:
>false [](start = 14, length = 5)

providing the name of this parameter would make it more descriptive

Reply:
I added helpers, that seems like the clearest approach.